### PR TITLE
[Reformer] Adapt Reformer MaskedLM Attn mask

### PR DESCRIPTION
**Pull request by patrickvonplaten**
The official trax code added an `attention_mask` to the LSH Attention layer a couple of weeks ago: https://github.com/google/trax/commit/94d7d8643d9a6ea38539486dea9ad16da30ec897#diff-a022408d1029c5cbeac49f4589f1b713R1180 . This PR adapts the Hugging Face code to have equal outputs to the official trax code for masked self attention.

Integration tests are adapted and run on branch: `reformer_trax_tests`.

### CODE REVIEW

File: src/transformers/modeling_reformer.py [hunk 1]
```diff
@@ -514,6 +514,15 @@ def _hash_vectors(self, vectors, num_hashes):
 
                 cur_product = cur_product * bucket_factor
 
+        if attention_mask is not None:
```
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:517]: If an `attention_mask` is given an extra bucket is added and buckets assignments for positions that are padded are given this new bucket. This code is new and thus changes the integration tests.

File: src/transformers/modeling_reformer.py [hunk 2]
```diff
@@ -669,45 +680,32 @@ def _attend(
...
         # Causal mask
         if self.is_decoder:
-            mask = torch.ge(query_indices.unsqueeze(-1), key_indices.unsqueeze(-2)).to(query_indices.device)
+            causal_mask = torch.ge(query_indices.unsqueeze(-1), key_indices.unsqueeze(-2)).to(query_indices.device)
 
         # Attention mask: chunk, look up correct mask value from key_value_bucket_idx
         # IMPORTANT: official trax code does not use a mask for LSH Atttention. Not sure why.
         if attention_mask is not None:
             # if chunked attention, the attention mask has to correspond to LSH order
+            attention_mask = attention_mask.to(torch.uint8)[:, None, :]
```
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:694]: cleaner code

File: src/transformers/modeling_reformer.py [hunk 3]
```diff
@@ -923,32 +921,29 @@ def forward(self, hidden_states, attention_mask=None, head_mask=None, output_att
...
         # chunk attention mask and look before and after
         if attention_mask is not None:
             attention_mask = attention_mask.to(torch.uint8)[:, None, :]
 
             if self.chunk_length < sequence_length:
                 attention_mask = self._split_seq_length_dim_to(attention_mask, -1, self.chunk_length, 1)
-                attention_mask_key = self._look_adjacent(attention_mask, self.num_chunks_before, self.num_chunks_after)
-            else:
-                attention_mask_key = attention_mask
+                attention_mask = self._look_adjacent(attention_mask, self.num_chunks_before, self.num_chunks_after)
```
Comment (patrickvonplaten) [inline src/transformers/modeling_reformer.py:932]: cleaner code

File: tests/test_modeling_reformer.py [hunk 4]
```diff
@@ -940,7 +940,7 @@ def test_lm_model_forward(self):
         hidden_states = model(input_ids=input_ids, attention_mask=attn_mask)[0]
         output_slice = hidden_states[1, -1, :5]
         expected_output_slice = torch.tensor(
-            [0.0324, -0.0121, 0.0615, 0.0031, -0.0297], dtype=torch.float, device=torch_device,
+            [0.0256, -0.0121, 0.0636, 0.0024, -0.0393], dtype=torch.float, device=torch_device,
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:943]: changed because of new LSH attention mask logic.

**Comment by patrickvonplaten** [comment]
Pinging @LysandreJik @thomwolf @flozi00 for notification.

File: tests/test_modeling_reformer.py [hunk 5]
```diff
@@ -407,7 +407,8 @@ def create_and_check_reformer_model_fp16_generate(self, config, input_ids, input
         model.to(torch_device)
         model.half()
         model.eval()
-        output = model.generate(input_ids, attention_mask=input_mask, do_sample=False)
+        # only use last 10 inputs for generation
+        output = model.generate(input_ids[:, -10:], attention_mask=input_mask, do_sample=False)
```
Comment (patrickvonplaten) [inline tests/test_modeling_reformer.py:411]: new assert in generate prevents to generate "less" than `max_length`.
